### PR TITLE
Porting changes from localStorage clear PR and adding unit tests

### LINKED ISSUE #15: Port the localStorage clearing feature from FSP
https://github.com/FullScreenShenanigans/FullScreenPokemon/pull/281/files

### PR DESCRIPTION
**Pull request by gpnoel**
closes #15

### CODE REVIEW

File: src/ItemValue.ts [hunk 1]
```diff
@@ -132,7 +132,7 @@ export class ItemValue implements IItemValue {
                 this.update();
             } else {
                 // Otherwise save the new version to memory
-                this.updateLocalStorage();
+                this.updateLocalStorage(true);
```
Comment (JoshuaKGoldberg) [inline src/ItemValue.ts]: Why?

File: src/ItemValue.ts [hunk 2]
```diff
@@ -155,7 +155,7 @@ export class ItemValue implements IItemValue {
      * @param value   The desired value to now store.
      */
     public setValue(value: any): void {
-        if (this.transformSet) {
+        if (this.hasOwnProperty("transformSet")) {
```
Comment (JoshuaKGoldberg) [inline src/ItemValue.ts]: Why?

File: src/ItemValue.ts [hunk 3]
```diff
@@ -264,10 +264,10 @@ export class ItemValue implements IItemValue {
      * 
      * @returns {Mixed}
      */
-    private retrieveLocalStorage(): void {
-        const value: any = localStorage.getItem(this.ItemsHolder.getPrefix() + this.key);
+    private retrieveLocalStorage(): any {
+        const value: any = this.ItemsHolder.getLocalStorage()[this.ItemsHolder.getPrefix() + this.key];
 
-        if (value === "undefined") {
+        if (!value) {
```
Comment (JoshuaKGoldberg) [inline src/ItemValue.ts]: You should add unit tests for values like `undefined` and `""` and `0`.

File: src/ItemsHoldr.ts [hunk 4]
```diff
@@ -302,11 +301,11 @@ export class ItemsHoldr implements IItemsHoldr {
     }
 
     /**
-     * Increases the value for the ItemValue under the given key, via addition for
+     * Decreases the value for the ItemValue under the given key, via addition for
```
Comment (JoshuaKGoldberg) [inline src/ItemsHoldr.ts]: Whoops...!

File: test/unit/ItemValueTests/checkModularity.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
+define(["mocks"], function (mocks) {
+    return function () {
+        var expect = require("chai").expect;
+
+        it("calls the modular function the correct number of times", function () {
+            var item = mocks.mockItemValue(mocks.mockItemsHoldr(), "weight", {
```
Comment (JoshuaKGoldberg) [inline test/unit/ItemValueTests/checkModularity.js]: For all of these, can you add explicit `// Arrange`/`// Act`/`// Assert` comments?

File: src/ItemValue.ts (hunk 1 above)
Comment (gpnoel) [inline src/ItemValue.ts]: If nothing is passed in (or autoSave isn't enabled), the value will not save to localStorage

File: src/ItemValue.ts (hunk 2 above)
Comment (gpnoel) [inline src/ItemValue.ts]: When I did make this change, the places in the file I looked at were checking with `hasOwnProperty` but after just doing another look in the file, `if (this. ...)` is used evenly with `hasOwnProperty`. Should I try to convert to `hasOwnProperty` for checks like these (except for boolean values ofc)  or just revert this change?

File: src/ItemValue.ts (hunk 1 above)
Comment (JoshuaKGoldberg) [inline src/ItemValue.ts]: I believe that's the intended purpose. If you create a new `ItemsHoldr` with no autosave and defaults defined in settings, it shouldn't write them to localStorage before manual saving is requested.

File: src/ItemValue.ts (hunk 2 above)
Comment (JoshuaKGoldberg) [inline src/ItemValue.ts]: Just revert the change. The `.hasOwnProperty(foo)` vs `.foo` debate is one we shouldn't wade into, largely because I haven't made up a solidified opinion for FSS projects.

File: src/ItemValue.ts (hunk 1 above)
Comment (gpnoel) [inline src/ItemValue.ts]: Hmm, ok.

File: test/unit/mocks.js [hunk 6]
```diff
@@ -1,18 +1,21 @@
...
 
     var mocks = {
         /**
-         * 
+         * @param [settings]   Settings for the ItemValue.
+         * @returns An ItemValue instance.
          */
-        mockItemValue: function (settings) {
-            return new ItemValue(settings)
+        mockItemValue: function (IH, key, settings) {
```
Comment (JoshuaKGoldberg) [inline test/unit/mocks.js]: Can you update the JSDoc for this? I'm also wary of names like `IH`; better to spell it out.

Comment (gpnoel) [inline test/unit/mocks.js]: Yeah, I remembered earlier today that I forgot to change that

File: src/ItemValue.ts (hunk 3 above)
Comment (JoshuaKGoldberg) [inline src/ItemValue.ts]: Re-thinking this change: if value is `0` then this will return undefined. Should revert the change on this line.

Comment (gpnoel) [inline src/ItemValue.ts]: Yeah, I saw this as a typo and tried doing `typeof value === "undefined"`, but I got an error and couldn't figure it out. So I changed it to `!value`. Just tried `typeof value === "undefined"` and all tests passed so I'm changing it to that.
